runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/RedistributeByKeyOverrideFactory.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.dataflow;
+
+import java.util.Collections;
+import org.apache.beam.sdk.runners.AppliedPTransform;
+import org.apache.beam.sdk.runners.PTransformOverrideFactory;
+import org.apache.beam.sdk.runners.PTransformOverrideFactory.PTransformReplacement;
+import org.apache.beam.sdk.transforms.DataflowGroupByKey;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.DoFn.Element;
+import org.apache.beam.sdk.transforms.DoFn.OutputReceiver;
+import org.apache.beam.sdk.transforms.DoFn.ProcessElement;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.Redistribute.RedistributeByKey;
+import org.apache.beam.sdk.transforms.Reify;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.sdk.transforms.windowing.ReshuffleTrigger;
+import org.apache.beam.sdk.transforms.windowing.TimestampCombiner;
+import org.apache.beam.sdk.transforms.windowing.Window;
+import org.apache.beam.sdk.util.IdentityWindowFn;
+import org.apache.beam.sdk.util.construction.PTransformReplacements;
+import org.apache.beam.sdk.util.construction.SingleInputOutputOverrideFactory;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.ValueInSingleWindow;
+import org.apache.beam.sdk.values.WindowingStrategy;
+import org.joda.time.Duration;
+
+@SuppressWarnings({
+  "nullness" // TODO(https://github.com/apache/beam/issues/20497)
```

Comment:
don't suppress in new files

Reply:
Then reified.apply(
              originalTransform.getAllowDuplicates()
                  ? DataflowGroupByKey.createWithAllowDuplicates()
                  : DataflowGroupByKey.create());
will cause check failure due to imcompatible types for .apply method.

Reply:
What are the incompatible types? Basically, treat nullness the same way you would treat any other type error. You can suppress at a very fine local level if absolutely necessary. But in this case I think you just need to add a null check if there is a nullable value probably.

Reply:
Moved the suppress to the corresponding place because I still can't find a way to resolve the incompatible types:

found   : @Initialized @NonNull PCollection<@Initialized @NonNull KV<K extends @Initialized @Nullable Object, @Initialized @NonNull Iterable<@Initialized @Nullable ValueInSingleWindow<V extends @Initialized @Nullable Object>>>>
  required: @UnknownInitialization @Nullable PCollection<@Initialized @NonNull KV<K extends @Initialized @Nullable Object, @Initialized @NonNull Iterable<@Initialized @NonNull ValueInSingleWindow<V extends @Initialized @Nullable Object>>>>

Reply:
Can you paste the whole message in a code block to make sure I am reading right? It looks like you are passing something where the items in the iterable are `@nullable ValueInSingleWindow` when it requires `@nonnull ValueInSingleWindow` so that would be an actual error.

Reply:
```

found   : @Initialized @NonNull PCollection<@Initialized @NonNull KV<K extends @Initialized @Nullable Object, @Initialized @NonNull Iterable<@Initialized @Nullable ValueInSingleWindow<V extends @Initialized @Nullable Object>>>>
742
  required: @UnknownInitialization @Nullable PCollection<@Initialized @NonNull KV<K extends @Initialized @Nullable Object, @Initialized @NonNull Iterable<@Initialized @NonNull ValueInSingleWindow<V extends @Initialized @Nullable Object>>>>
743
/runner/_work/beam/beam/runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/RedistributeByKeyOverrideFactory.java:89: error: [argument] incompatible argument for parameter t of PCollection.apply.
744
                  ? DataflowGroupByKey.createWithAllowDuplicates()
```